docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,350 @@
+ * @param[in] addr, desired page aligned start address, NULL if no desired address
+ * @param[in] length, size of the region in multiples of page size in bytes
+ * @param[in] flags, must be MAP_SHARED | MAP_FIXED masked with optional flags:
+ *              MAP_POPULATE: hint for kernel to EAUG pages as soon as possible.
+ *              MAP_GROWSDOWN: used for stacks. The mapping will grow down to the next mapping.
+ * @param[out] out_addr, start address of region reserved by OS
+ * @return 0 on success, or error code similar to mmap
+ */
+int mmap_ocall(void* addr, size_t length, int flags, void** out_addr);
+/**
+ *
+ * call OS mprotect to change permission, type, or notify EACCEPT done after TRIM
+ * 
+ * @param[in] addr, start address of the memory to change protections
+ * @param[in] length, length of the area.  This must be a multiple of the page size.
+ * @param[in] prot, this must be OR'ed of following:
+ *            PROT_READ
+ *            PROT_WRITE
+ *            PROT_EXEC
+ *            PROT_TRIM: change the page type to PT_TRIM, implies RW. 
+ *            PROT_TCS: change the page type to PT_TCS
+ *            PROT_SS_FIRST
+ *            PROT_SS_REST
+ *            PROT_NONE: Signal the kernel EACCEPT is done for PT_TRIM pages. 
+ * @return 0 on success, or error code similar to mprotect
+ */
+
+mprotect_ocall(void*addr, int flags)
+
+```
+
+Metadata,  File format
+---------------------------------------
+
+Enclave metadata and file format are runtime specific. Detailed design is out of scope of this document.
+It is required that the enclave file should include metadata of memory layout of initial code and data (e.g., program headers and PT_LOAD segments in ELF file), any reserved region for special purposes, e.g. minimal heap, stack, TCS areas, SSAs for expected minimal number of threads, etc.
+ 
+
+EMM Initialization and Tracking Initially Committed (EADDed) Regions
+---------------------------------------------------------------------
+
+The memory manager must be initialized in first ECALL (ECMD_INIT_ENCLAVE in Intel SGX SDK) before any other clients can use it. Therefore, code and data of the memory manager will be part of initial enclave image that are loaded with EADD before EINIT, and as a part of the trusted runtime. 
+
+The trusted runtime should enumerate all initial committed regions (code, data, heap, stack, TCS, SSA), and call the EMM internal APIs to setup initial entries in the EMA list to track existing regions and mark some of them not modifiable by EMM public APIs. Runtime also ensures there is enough reserved space on heap for EMM to create the initial EMA list and the entries. Once initialized, the memory manager can reserve its own space for future expansion of the EMA list, and special EMAs to hold EMA objects. To keep it simple, the expansion can be done eagerly: commit more pages for EMA list once unused committed space in the EMA List Region below certain threshold.
```

Comment:
Note for OE: This option indicates some work to the OE loader. Currently we don't explicitly pass the full enclave layout information (only necessary ones for relocations and heap allocators) to the enclave.

Reply:
Please see updated version with abstraction layer spec